pkg/gadgets/internal/socketenricher/bpf/sockets-map.bpf.c

```diff
@@ -25,25 +25,111 @@ struct {
 	__type(value, void *);
 } start SEC(".maps");
 
+const volatile __u64 socket_file_ops_addr = 0;
+
 static __always_inline void
-update_sockets_map(struct sock *sock)
+prepare_socket_key(struct sockets_key *socket_key, struct sock *sock)
 {
-	struct sockets_key socket_key = {0,};
 	struct inet_sock *inet_sock = (struct inet_sock *)sock;
-	BPF_CORE_READ_INTO(&socket_key.netns, sock, __sk_common.skc_net.net, ns.inum);
-	BPF_CORE_READ_INTO(&socket_key.family, sock, __sk_common.skc_family);
-	socket_key.proto = BPF_CORE_READ_BITFIELD_PROBED(sock, sk_protocol);
-	socket_key.port = bpf_ntohs(BPF_CORE_READ(inet_sock, inet_sport));
+	BPF_CORE_READ_INTO(&socket_key->netns, sock, __sk_common.skc_net.net, ns.inum);
+	BPF_CORE_READ_INTO(&socket_key->family, sock, __sk_common.skc_family);
+	socket_key->proto = BPF_CORE_READ_BITFIELD_PROBED(sock, sk_protocol);
+	socket_key->port = bpf_ntohs(BPF_CORE_READ(inet_sock, inet_sport));
+}
+
+static __always_inline void
+insert_current_socket(struct sock *sock)
+{
+	struct sockets_key socket_key = {0,};
+	prepare_socket_key(&socket_key, sock);
 
 	struct sockets_value socket_value = {0,};
+	// use 'current' task
 	struct task_struct *task = (struct task_struct*) bpf_get_current_task();
 	socket_value.mntns = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
 	socket_value.pid_tgid = bpf_get_current_pid_tgid();
 	bpf_get_current_comm(&socket_value.task, sizeof(socket_value.task));
 	socket_value.sock = (__u64) sock;
+
 	bpf_map_update_elem(&sockets, &socket_key, &socket_value, BPF_ANY);
 }
 
+static __always_inline void
+insert_socket_from_iter(struct sock *sock, struct task_struct *task)
+{
+	struct sockets_key socket_key = {0,};
+	prepare_socket_key(&socket_key, sock);
+
+	struct sockets_value socket_value = {0,};
+	// use given task
+	socket_value.pid_tgid = ((u64)task->tgid) << 32 | task->pid;
```

Comment:
I tried to remove BPF_CORE_READ everywhere for iterator, but the following ones failed:
- `struct sock *sock = BPF_CORE_READ(socket, sk);`
- `__u16 family = BPF_CORE_READ(sock, __sk_common.skc_family);`
- `if (BPF_CORE_READ(sock, __sk_common.skc_state) == TCP_LISTEN)`

Example:

```
// file->private_data is a struct socket because we checked f_op.
struct socket *socket = (struct socket *) file->private_data;
struct sock *sock = socket->sk;
```

```
; struct socket *socket = (struct socket *) file->private_data;
11: (79) r1 = *(u64 *)(r1 +200)       ; R1_w=scalar()
; struct sock *sock = socket->sk;
12: (79) r7 = *(u64 *)(r1 +24)
R1 invalid mem access 'scalar'

processed 12 insns (limit 1000000) max_states_per_insn 0 total_states 1 peak_states 1 mark_read 1
Error: running gadget: running gadget: installing tracer: failed to load ebpf program: field
IgSocketsIt: program ig_sockets_it: load program: permission denied: 12: (79) r7 = *(u64 *)(r1 +24):
R1 invalid mem access 'scalar' (24 line(s) omitted)
```

Printing errors with:
```
	if err := spec.LoadAndAssign(&se.objs, nil); err != nil {
		var errVerifier *ebpf.VerifierError
		if errors.As(err, &errVerifier) {
			fmt.Printf("loading ebpf program:\n%+v\n", errVerifier)
		}
		return fmt.Errorf("failed to load ebpf program: %w", err)
	}
```

It seems to work for the structs that were passed as arguments to the iterator (`struct bpf_iter__task_file`) such as `struct task_struct`, `struct file`... but it does not work after the cast via `(struct socket *) file->private_data`.